src/main/java/io/cdap/delta/bigquery/MultiGCSWriter.java

```diff
@@ -250,16 +259,40 @@ private Schema getTargetSchema(Schema schema) {
           name (string)
           _before_id (long)
           _before_name (string)
+
+        If the source used in the pipeline is generating UN-ORDERED events, then for the same example above,
+        staging table would contain following columns:
+
+          _op (string)
+          _batch_id (long)
+          _sequence_num (long)
+          _source_timestamp (long)
+          _row_id (string)
+          id (long)
+          name (string)
      */
     private Schema getStagingSchema(Schema schema) {
-      List<Schema.Field> fields = new ArrayList<>(2 * schema.getFields().size() + 3);
-      fields.add(Schema.Field.of("_op", Schema.of(Schema.Type.STRING)));
-      fields.add(Schema.Field.of("_batch_id", Schema.of(Schema.Type.LONG)));
+      int fieldLength = sourceEventOrdering == SourceProperties.Ordering.ORDERED ?
+        2 * schema.getFields().size() + 3 // (source schema + _before_ columns) + _op, _batch_id, _sequence_num
+        : schema.getFields().size() + 4; // source schema fields + _op, _batch_id, _sequence_num, _row_id
+
+      List<Schema.Field> fields = new ArrayList<>(fieldLength);
+
+      fields.add(Schema.Field.of(Constants.OPERATION, Schema.of(Schema.Type.STRING)));
+      fields.add(Schema.Field.of(Constants.BATCH_ID, Schema.of(Schema.Type.LONG)));
+      fields.add(Schema.Field.of(Constants.SEQUENCE_NUM, Schema.of(Schema.Type.LONG)));
+
+      // add all fields from source schema
       fields.addAll(schema.getFields());
+
+      if (sourceEventOrdering == SourceProperties.Ordering.UN_ORDERED) {
+        // add _row_id and _source_timestamp fields to handle un-ordered events
+        fields.add(Schema.Field.of(Constants.ROW_ID, Schema.of(Schema.Type.STRING)));
+        return Schema.recordOf(schema.getRecordName() + ".staging", fields);
```

Comment:
nit: 
How about we do :

if (unordered ) {
   add row_id filed
} else {
   add before fileds
}
return Schema.recordOf(schema.getRecordName() + ".staging", fields);

this will make the code more readable and avoid the duplicate of the "return" statement

Reply:
I agree in general about having single return statements for method for readability, however for cases if we can simplify nesting would be better to  return early. In this case it adds one level more nesting for `else` which is probably ok. updated the code.